### PR TITLE
Add support for non-json reponse result to http.send builtin.

### LINKED ISSUE #1258: http.send fails to yield a response body if it is not JSON
# Expected Behavior

The [language reference](https://www.openpolicyagent.org/docs/language-reference.html#built-in-functions) describes `http.send` as simply returning the response body as a component of `output`. It sounds from the documentation that this `body` member ought to be a string, which could then be further processed with `json.unmarshal` if it's JSON, or otherwise if it's not.

# Actual Behavior

If the HTTP server response is parsable as JSON, that appears in `output.body` as a term, i.e. unmarshaled from a string. If the HTTP server responds with a body that doesn't contain valid JSON, what we get is `null`. The `output` has nothing to indicate what happened to the body.

See https://github.com/open-policy-agent/opa/blob/0d9cd4a89c34b062e9fcc1e98214f4ea8392e00f/topdown/http.go#L275 for logic that unconditionally tries to decode the string as JSON.

# Steps to Reproduce the Problem

Invoke `http.send` against a `url` that will return a non-JSON response payload.

# Additional Info

If nothing else, the documentation should describe the current state of affairs. This would make it easier to write Rego using `http.send` since it would be more obvious what can be done with the `output`, and which kinds of responses can be handled in general.

For speaking to non-JSON endpoints, one possibility is to add a `raw_body` member to `output`. This would preserve the existing behavior while also providing a more general capability for non-JSON data.

Another would be to fall back to `body` containing the raw response body if JSON parsing fails, or if the `Content-type` header doesn't indicate JSON. But it is probably confusing if `body` is sometimes parsed JSON and sometimes a raw string, and the `Content-type` is unfortunately not always reliable in the wild.

The idea in https://github.com/open-policy-agent/opa/issues/1176#issuecomment-456191566 also seems interesting in the longer term, replacing the low-level `http.send` facility with a higher-level abstraction over remote data sources.

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Fixes #1258

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: docs/book/language-reference.md [hunk 1]
```diff
@@ -209,9 +209,10 @@ evaluation query will always return the same value.
...
...method``, ``url`` and  optionally ``body``, ``enable_redirect``, ``disable_json_decode``, ``headers``, ``tls_use_system_certs``, ``tls_ca_cert_file``, ``tls_ca_cert_env_variable``, ``tls_client_cert_env_variable``, ``tls_client_key_env_variable`` or ``tls_client_cert_file``, ``tls_client_key_file`` . For example, ``http.send({"method": "get", "url": "http://www.openpolicyagent.org/", "headers": {"X-Foo":"bar", "X-Opa": "rules"}}, output)``. ``output`` is an object containing keys ``status``, ``status_code``, ``body`` and ``raw_body`` which represent the HTTP status, status code, JSON value from the response body and response body as string respectively. Sample output, ``{"status": "200 OK", "status_code": 200, "body": {"hello": "world"}, "raw_body": "{\"hello\": \"world\ "}"``}. By default, http redirects are not enabled. To enable, set ``enable_redirect`` to ``true``. Also ``disable_json_decode`` is set to ``true`` by default. This means if the HTTP server response does not specify the ``Content-type`` as ``application/json``, the response body will not be JSON decoded ie. output's ``body`` field will be ``nil``. To change this behaviour, set ``disable_json_decode`` to ``false``.|
```
Comment (agurney) [inline docs/book/language-reference.md]: I like the logic of always returning a `raw_body`, and if possible returning a JSON `body` - that helps with the common existing case of JSON responses, while allowing more straightforward code in other cases (compared to conditionally switching between having either a `body` or `raw_body`).

The `disable_json_decode` flag seems a bit hard to understand. I find disable-if-true to be confusing in general. If we rephrase this as `force_json_decode`, which is false by default, then we get the behavior:
1. Responses claiming to be `application/json` will always be JSON-decoded into `body`, if possible.
2. Other responses will not be so decoded, _unless_ `force_json_decode` is set.

File: topdown/http.go [hunk 2]
```diff
@@ -272,12 +281,24 @@ func executeHTTPRequest(bctx BuiltinContext, obj ast.Object) (ast.Value, error)
...
+	var resultRawBody []byte
+
+	var buf bytes.Buffer
+	tee := io.TeeReader(resp.Body, &buf)
+	resultRawBody, err = ioutil.ReadAll(tee)
+	if err != nil {
+		return nil, err
+	}
+
+	if isContentTypeJSON(resp.Header) || (!isContentTypeJSON(resp.Header) && !disableJSONDecode) {
```
Comment (agurney) [inline topdown/http.go]: suggest `if (isContentTypeJSON(resp.Header) || forceJSONDecode) { ... }`

File: docs/book/language-reference.md (hunk 1 above)
Comment (tsandall) [reply docs/book/language-reference.md]: @agurney I tend to agree however the only downside is that change would be backwards incompatible (i.e., users that had integrated their policy with an endpoint that doesn't set Content-Type properly would be broken.)

That said, now (if any) is the time to make backwards incompatible changes.

**Comment by ashutosh-narkar** [comment]
A new field `force_json_decode ` has been added to the `http.send` command. The behavior is similar to what @agurney suggested.

File: docs/book/language-reference.md [hunk 3]
```diff
@@ -209,9 +209,10 @@ evaluation query will always return the same value.
...
 
 ### HTTP
-| Built-in | Description |
-| ------- |-------------|
-| <span class="opa-keep-it-together">``output := http.send(request)``</span> | ``http.send`` executes a HTTP request and returns the response.``request`` is an object containing keys ``method``, ``url`` and  optionally ``body``, ``enable_redirect``, ``headers``, ``tls_use_system_certs``, ``tls_ca_cert_file``, ``tls_ca_cert_env_variable``, ``tls_client_cert_env_variable``, ``tls_client_key_env_variable`` or ``tls_client_cert_file``, ``tls_client_key_file`` . For example, ``http.send({"method": "get", "url": "http://www.openpolicyagent.org/", "headers": {"X-Foo":"bar", "X-Opa": "rules"}}, output)``. ``output`` is an object containing keys ``status``, ``status_code`` and ``body`` which represent the HTTP status, status code and response body respectively. Sample output, ``{"status": "200 OK", "status_code": 200, "body": null``}. By default, http redirects are not enabled. To enable, set ``enable_redirect`` to ``true``.|
+
+| Built-in | Inputs | Description |
```
Comment (tsandall) [inline docs/book/language-reference.md]: Nit: It looks like the merge re-added the Inputs column that we got rid of in one of the recent docs improvement PRs. Please scrap this column and fix the built-in call format to match the others.

File: docs/book/language-reference.md [hunk 4]
```diff
@@ -209,9 +209,10 @@ evaluation query will always return the same value.
...
...eys ``method``, ``url`` and  optionally ``body``, ``enable_redirect``, ``force_json_decode``, ``headers``, ``tls_use_system_certs``, ``tls_ca_cert_file``, ``tls_ca_cert_env_variable``, ``tls_client_cert_env_variable``, ``tls_client_key_env_variable`` or ``tls_client_cert_file``, ``tls_client_key_file`` . For example, ``http.send({"method": "get", "url": "http://www.openpolicyagent.org/", "headers": {"X-Foo":"bar", "X-Opa": "rules"}}, output)``. ``output`` is an object containing keys ``status``, ``status_code``, ``body`` and ``raw_body`` which represent the HTTP status, status code, JSON value from the response body and response body as string respectively. Sample output, ``{"status": "200 OK", "status_code": 200, "body": {"hello": "world"}, "raw_body": "{\"hello\": \"world\ "}"``}. By default, http redirects are not enabled. To enable, set ``enable_redirect`` to ``true``. Also ``force_json_decode`` is set to ``false`` by default. This means if the HTTP server response does not specify the ``Content-type`` as ``application/json``, the response body will not be JSON decoded ie. output's ``body`` field will be ``nil``. To change this behaviour, set ``force_json_decode`` to ``true``.|
```
Comment (tsandall) [inline docs/book/language-reference.md]: Nit: `By default, [http->HTTP] redirects`.

File: topdown/http.go [hunk 5]
```diff
@@ -272,12 +281,24 @@ func executeHTTPRequest(bctx BuiltinContext, obj ast.Object) (ast.Value, error)
...
+
+	var buf bytes.Buffer
+	tee := io.TeeReader(resp.Body, &buf)
+	resultRawBody, err = ioutil.ReadAll(tee)
+	if err != nil {
+		return nil, err
+	}
+
+	if isContentTypeJSON(resp.Header) || forceJSONDecode {
+		json.NewDecoder(&buf).Decode(&resultBody)
```
Comment (tsandall) [inline topdown/http.go]: What about errors?

File: topdown/http.go [hunk 6]
```diff
@@ -272,12 +281,24 @@ func executeHTTPRequest(bctx BuiltinContext, obj ast.Object) (ast.Value, error)
...
+	}
 
 	result := make(map[string]interface{})
 	result["status"] = resp.Status
 	result["status_code"] = resp.StatusCode
 	result["body"] = resultBody
+	result["raw_body"] = string(resultRawBody)
```
Comment (tsandall) [inline topdown/http.go]: Out of curiosity, what happens if the raw body contains binary characters?

File: docs/book/language-reference.md (hunk 3 above)
Comment (ashutosh-narkar) [reply docs/book/language-reference.md]: Done.

File: docs/book/language-reference.md (hunk 4 above)
Comment (ashutosh-narkar) [reply docs/book/language-reference.md]: Done.

File: topdown/http.go (hunk 5 above)
Comment (ashutosh-narkar) [reply topdown/http.go]: Added a note.

File: topdown/http.go (hunk 6 above)
Comment (ashutosh-narkar) [reply topdown/http.go]: If the response body contains binary characters, the raw body will have them too.
